{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "43710102-d536-4d85-a8cc-f29a117ff646",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.0     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(rtracklayer)\n",
    "library(tidyverse)\n",
    "library(ChrAccR)\n",
    "library(motifmatchr)\n",
    "library(SummarizedExperiment)\n",
    "library(ggpubr)\n",
    "library(patchwork)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f24e2223-9715-4ccc-a186-94142aeef949",
   "metadata": {},
   "source": [
    "### read all TF data: cpm, motif matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15fe57a1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "deseq_dir <- \"../../output/02-atac/04/deseq_data_Plate_TF_Dose/\"\n",
    "plotdir <- \"../../plots/02-atac/17/\"\n",
    "outdir <- \"../../output/02-atac/17/\"\n",
    "\n",
    "dir.create(plotdir, showWarnings = F, recursive=T)\n",
    "dir.create(outdir, showWarnings = F, recursive=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5bef7f1-0c82-467a-9acf-6cc5375bccbd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 252</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X1_HEK293T_P1A2_SP4_d100</th><th scope=col>X2_HEK293T_P1B2_SP4_d100</th><th scope=col>X3_HEK293T_P1C2_ELF1_d100</th><th scope=col>X4_HEK293T_P1D2_ELF1_d100</th><th scope=col>X5_HEK293T_P1E2_FOXP1_d100</th><th scope=col>X6_HEK293T_P1F2_FOXP1_d100</th><th scope=col>X7_HEK293T_P1G2_GATA1_d100</th><th scope=col>X8_HEK293T_P1H2_GATA1_d100</th><th scope=col>X9_HEK293T_P1A3_SP4_d075</th><th scope=col>X10_HEK293T_P1B3_SP4_d075</th><th scope=col>⋯</th><th scope=col>X407_HEK293T_P3G11_SPI1_d005</th><th scope=col>X408_HEK293T_P3H11_SPI1_d005</th><th scope=col>X409_HEK293T_P3A12_GFP_d100</th><th scope=col>X410_HEK293T_P3B12_GFP_d100</th><th scope=col>X411_HEK293T_P3C12_tetCTCF_d100</th><th scope=col>X412_HEK293T_P3D12_tetCTCF_d100</th><th scope=col>X413_HEK293T_P3E12_GFP_d100</th><th scope=col>X414_HEK293T_P3F12_GFP_d100</th><th scope=col>X415_HEK293T_P3G12_negctl_d100</th><th scope=col>X416_HEK293T_P3H12_negctl_d100</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>HEK293T_P3A12_GFP_d100_peak_1</th><td>1.588078</td><td>0.8776116</td><td>0.8770458</td><td>1.688634</td><td>2.175640</td><td> 1.393018</td><td>1.558472</td><td>1.6576188</td><td>1.483490</td><td>0.425564</td><td>⋯</td><td>1.1968762</td><td>1.770908</td><td>2.7855153</td><td>1.8936426</td><td>1.1304493</td><td>1.8257871</td><td>0.7031904</td><td>1.0822624</td><td>2.313741</td><td>1.4081195</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3A12_GFP_d100_peak_17</th><td>6.087634</td><td>8.9955187</td><td>8.9897199</td><td>9.118624</td><td>8.857964</td><td>10.795892</td><td>8.500757</td><td>8.6564539</td><td>8.900937</td><td>5.957896</td><td>⋯</td><td>9.1760505</td><td>7.083633</td><td>9.2850511</td><td>7.5745705</td><td>7.3479207</td><td>7.9117439</td><td>9.1414749</td><td>6.8543286</td><td>7.519658</td><td>8.4487170</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3G8_SPI1_d075_peak_6</th><td>1.588078</td><td>3.0716405</td><td>1.0963073</td><td>1.688634</td><td>2.175640</td><td> 2.263655</td><td>0.991755</td><td>2.0259786</td><td>1.483490</td><td>1.276692</td><td>⋯</td><td>0.7979174</td><td>2.361211</td><td>4.6425255</td><td>2.3670533</td><td>1.6956740</td><td>0.6085957</td><td>2.8127615</td><td>0.3607541</td><td>1.156870</td><td>0.7040597</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P1B6_SP4_d005_peak_3</th><td>1.323399</td><td>1.7552232</td><td>2.4118761</td><td>3.714995</td><td>3.263461</td><td> 3.134291</td><td>3.116944</td><td>2.2101584</td><td>2.472483</td><td>1.276692</td><td>⋯</td><td>1.1968762</td><td>1.770908</td><td>2.7855153</td><td>0.9468213</td><td>0.5652247</td><td>2.4343827</td><td>1.4063807</td><td>2.1645248</td><td>4.627482</td><td>3.5202987</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3G7_SPI1_d100_peak_10</th><td>2.382118</td><td>0.8776116</td><td>2.8503990</td><td>1.350907</td><td>2.020237</td><td> 1.393018</td><td>1.416793</td><td>0.3683597</td><td>0.988993</td><td>1.702256</td><td>⋯</td><td>2.7927110</td><td>2.361211</td><td>0.9285051</td><td>1.4202320</td><td>0.5652247</td><td>1.2171914</td><td>       NA</td><td>1.4430165</td><td>1.156870</td><td>1.4081195</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3F10_SPI1_d025_peak_18</th><td>2.117438</td><td>1.7552232</td><td>3.2889219</td><td>3.039541</td><td>2.952655</td><td> 3.482546</td><td>3.967020</td><td>4.2361370</td><td>3.461476</td><td>2.553384</td><td>⋯</td><td>2.7927110</td><td>1.180605</td><td>3.7140204</td><td>5.2075172</td><td>0.5652247</td><td>1.8257871</td><td>0.7031904</td><td>4.3290496</td><td>      NA</td><td>0.7040597</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 252\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & X1\\_HEK293T\\_P1A2\\_SP4\\_d100 & X2\\_HEK293T\\_P1B2\\_SP4\\_d100 & X3\\_HEK293T\\_P1C2\\_ELF1\\_d100 & X4\\_HEK293T\\_P1D2\\_ELF1\\_d100 & X5\\_HEK293T\\_P1E2\\_FOXP1\\_d100 & X6\\_HEK293T\\_P1F2\\_FOXP1\\_d100 & X7\\_HEK293T\\_P1G2\\_GATA1\\_d100 & X8\\_HEK293T\\_P1H2\\_GATA1\\_d100 & X9\\_HEK293T\\_P1A3\\_SP4\\_d075 & X10\\_HEK293T\\_P1B3\\_SP4\\_d075 & ⋯ & X407\\_HEK293T\\_P3G11\\_SPI1\\_d005 & X408\\_HEK293T\\_P3H11\\_SPI1\\_d005 & X409\\_HEK293T\\_P3A12\\_GFP\\_d100 & X410\\_HEK293T\\_P3B12\\_GFP\\_d100 & X411\\_HEK293T\\_P3C12\\_tetCTCF\\_d100 & X412\\_HEK293T\\_P3D12\\_tetCTCF\\_d100 & X413\\_HEK293T\\_P3E12\\_GFP\\_d100 & X414\\_HEK293T\\_P3F12\\_GFP\\_d100 & X415\\_HEK293T\\_P3G12\\_negctl\\_d100 & X416\\_HEK293T\\_P3H12\\_negctl\\_d100\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tHEK293T\\_P3A12\\_GFP\\_d100\\_peak\\_1 & 1.588078 & 0.8776116 & 0.8770458 & 1.688634 & 2.175640 &  1.393018 & 1.558472 & 1.6576188 & 1.483490 & 0.425564 & ⋯ & 1.1968762 & 1.770908 & 2.7855153 & 1.8936426 & 1.1304493 & 1.8257871 & 0.7031904 & 1.0822624 & 2.313741 & 1.4081195\\\\\n",
       "\tHEK293T\\_P3A12\\_GFP\\_d100\\_peak\\_17 & 6.087634 & 8.9955187 & 8.9897199 & 9.118624 & 8.857964 & 10.795892 & 8.500757 & 8.6564539 & 8.900937 & 5.957896 & ⋯ & 9.1760505 & 7.083633 & 9.2850511 & 7.5745705 & 7.3479207 & 7.9117439 & 9.1414749 & 6.8543286 & 7.519658 & 8.4487170\\\\\n",
       "\tHEK293T\\_P3G8\\_SPI1\\_d075\\_peak\\_6 & 1.588078 & 3.0716405 & 1.0963073 & 1.688634 & 2.175640 &  2.263655 & 0.991755 & 2.0259786 & 1.483490 & 1.276692 & ⋯ & 0.7979174 & 2.361211 & 4.6425255 & 2.3670533 & 1.6956740 & 0.6085957 & 2.8127615 & 0.3607541 & 1.156870 & 0.7040597\\\\\n",
       "\tHEK293T\\_P1B6\\_SP4\\_d005\\_peak\\_3 & 1.323399 & 1.7552232 & 2.4118761 & 3.714995 & 3.263461 &  3.134291 & 3.116944 & 2.2101584 & 2.472483 & 1.276692 & ⋯ & 1.1968762 & 1.770908 & 2.7855153 & 0.9468213 & 0.5652247 & 2.4343827 & 1.4063807 & 2.1645248 & 4.627482 & 3.5202987\\\\\n",
       "\tHEK293T\\_P3G7\\_SPI1\\_d100\\_peak\\_10 & 2.382118 & 0.8776116 & 2.8503990 & 1.350907 & 2.020237 &  1.393018 & 1.416793 & 0.3683597 & 0.988993 & 1.702256 & ⋯ & 2.7927110 & 2.361211 & 0.9285051 & 1.4202320 & 0.5652247 & 1.2171914 &        NA & 1.4430165 & 1.156870 & 1.4081195\\\\\n",
       "\tHEK293T\\_P3F10\\_SPI1\\_d025\\_peak\\_18 & 2.117438 & 1.7552232 & 3.2889219 & 3.039541 & 2.952655 &  3.482546 & 3.967020 & 4.2361370 & 3.461476 & 2.553384 & ⋯ & 2.7927110 & 1.180605 & 3.7140204 & 5.2075172 & 0.5652247 & 1.8257871 & 0.7031904 & 4.3290496 &       NA & 0.7040597\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 252\n",
       "\n",
       "| <!--/--> | X1_HEK293T_P1A2_SP4_d100 &lt;dbl&gt; | X2_HEK293T_P1B2_SP4_d100 &lt;dbl&gt; | X3_HEK293T_P1C2_ELF1_d100 &lt;dbl&gt; | X4_HEK293T_P1D2_ELF1_d100 &lt;dbl&gt; | X5_HEK293T_P1E2_FOXP1_d100 &lt;dbl&gt; | X6_HEK293T_P1F2_FOXP1_d100 &lt;dbl&gt; | X7_HEK293T_P1G2_GATA1_d100 &lt;dbl&gt; | X8_HEK293T_P1H2_GATA1_d100 &lt;dbl&gt; | X9_HEK293T_P1A3_SP4_d075 &lt;dbl&gt; | X10_HEK293T_P1B3_SP4_d075 &lt;dbl&gt; | ⋯ ⋯ | X407_HEK293T_P3G11_SPI1_d005 &lt;dbl&gt; | X408_HEK293T_P3H11_SPI1_d005 &lt;dbl&gt; | X409_HEK293T_P3A12_GFP_d100 &lt;dbl&gt; | X410_HEK293T_P3B12_GFP_d100 &lt;dbl&gt; | X411_HEK293T_P3C12_tetCTCF_d100 &lt;dbl&gt; | X412_HEK293T_P3D12_tetCTCF_d100 &lt;dbl&gt; | X413_HEK293T_P3E12_GFP_d100 &lt;dbl&gt; | X414_HEK293T_P3F12_GFP_d100 &lt;dbl&gt; | X415_HEK293T_P3G12_negctl_d100 &lt;dbl&gt; | X416_HEK293T_P3H12_negctl_d100 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HEK293T_P3A12_GFP_d100_peak_1 | 1.588078 | 0.8776116 | 0.8770458 | 1.688634 | 2.175640 |  1.393018 | 1.558472 | 1.6576188 | 1.483490 | 0.425564 | ⋯ | 1.1968762 | 1.770908 | 2.7855153 | 1.8936426 | 1.1304493 | 1.8257871 | 0.7031904 | 1.0822624 | 2.313741 | 1.4081195 |\n",
       "| HEK293T_P3A12_GFP_d100_peak_17 | 6.087634 | 8.9955187 | 8.9897199 | 9.118624 | 8.857964 | 10.795892 | 8.500757 | 8.6564539 | 8.900937 | 5.957896 | ⋯ | 9.1760505 | 7.083633 | 9.2850511 | 7.5745705 | 7.3479207 | 7.9117439 | 9.1414749 | 6.8543286 | 7.519658 | 8.4487170 |\n",
       "| HEK293T_P3G8_SPI1_d075_peak_6 | 1.588078 | 3.0716405 | 1.0963073 | 1.688634 | 2.175640 |  2.263655 | 0.991755 | 2.0259786 | 1.483490 | 1.276692 | ⋯ | 0.7979174 | 2.361211 | 4.6425255 | 2.3670533 | 1.6956740 | 0.6085957 | 2.8127615 | 0.3607541 | 1.156870 | 0.7040597 |\n",
       "| HEK293T_P1B6_SP4_d005_peak_3 | 1.323399 | 1.7552232 | 2.4118761 | 3.714995 | 3.263461 |  3.134291 | 3.116944 | 2.2101584 | 2.472483 | 1.276692 | ⋯ | 1.1968762 | 1.770908 | 2.7855153 | 0.9468213 | 0.5652247 | 2.4343827 | 1.4063807 | 2.1645248 | 4.627482 | 3.5202987 |\n",
       "| HEK293T_P3G7_SPI1_d100_peak_10 | 2.382118 | 0.8776116 | 2.8503990 | 1.350907 | 2.020237 |  1.393018 | 1.416793 | 0.3683597 | 0.988993 | 1.702256 | ⋯ | 2.7927110 | 2.361211 | 0.9285051 | 1.4202320 | 0.5652247 | 1.2171914 |        NA | 1.4430165 | 1.156870 | 1.4081195 |\n",
       "| HEK293T_P3F10_SPI1_d025_peak_18 | 2.117438 | 1.7552232 | 3.2889219 | 3.039541 | 2.952655 |  3.482546 | 3.967020 | 4.2361370 | 3.461476 | 2.553384 | ⋯ | 2.7927110 | 1.180605 | 3.7140204 | 5.2075172 | 0.5652247 | 1.8257871 | 0.7031904 | 4.3290496 |       NA | 0.7040597 |\n",
       "\n"
      ],
      "text/plain": [
       "                                X1_HEK293T_P1A2_SP4_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.588078                \n",
       "HEK293T_P3A12_GFP_d100_peak_17  6.087634                \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.588078                \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.323399                \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  2.382118                \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.117438                \n",
       "                                X2_HEK293T_P1B2_SP4_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.8776116               \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.9955187               \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   3.0716405               \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.7552232               \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.8776116               \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.7552232               \n",
       "                                X3_HEK293T_P1C2_ELF1_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.8770458                \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.9897199                \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.0963073                \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.4118761                \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  2.8503990                \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 3.2889219                \n",
       "                                X4_HEK293T_P1D2_ELF1_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.688634                 \n",
       "HEK293T_P3A12_GFP_d100_peak_17  9.118624                 \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.688634                 \n",
       "HEK293T_P1B6_SP4_d005_peak_3    3.714995                 \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.350907                 \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 3.039541                 \n",
       "                                X5_HEK293T_P1E2_FOXP1_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   2.175640                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.857964                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.175640                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    3.263461                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  2.020237                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.952655                  \n",
       "                                X6_HEK293T_P1F2_FOXP1_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1    1.393018                 \n",
       "HEK293T_P3A12_GFP_d100_peak_17  10.795892                 \n",
       "HEK293T_P3G8_SPI1_d075_peak_6    2.263655                 \n",
       "HEK293T_P1B6_SP4_d005_peak_3     3.134291                 \n",
       "HEK293T_P3G7_SPI1_d100_peak_10   1.393018                 \n",
       "HEK293T_P3F10_SPI1_d025_peak_18  3.482546                 \n",
       "                                X7_HEK293T_P1G2_GATA1_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.558472                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.500757                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   0.991755                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    3.116944                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.416793                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 3.967020                  \n",
       "                                X8_HEK293T_P1H2_GATA1_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.6576188                 \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.6564539                 \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.0259786                 \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.2101584                 \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.3683597                 \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 4.2361370                 \n",
       "                                X9_HEK293T_P1A3_SP4_d075\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.483490                \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.900937                \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.483490                \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.472483                \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.988993                \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 3.461476                \n",
       "                                X10_HEK293T_P1B3_SP4_d075 ⋯\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.425564                  ⋯\n",
       "HEK293T_P3A12_GFP_d100_peak_17  5.957896                  ⋯\n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.276692                  ⋯\n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.276692                  ⋯\n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.702256                  ⋯\n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.553384                  ⋯\n",
       "                                X407_HEK293T_P3G11_SPI1_d005\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.1968762                   \n",
       "HEK293T_P3A12_GFP_d100_peak_17  9.1760505                   \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   0.7979174                   \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.1968762                   \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  2.7927110                   \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.7927110                   \n",
       "                                X408_HEK293T_P3H11_SPI1_d005\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.770908                    \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.083633                    \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.361211                    \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.770908                    \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  2.361211                    \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.180605                    \n",
       "                                X409_HEK293T_P3A12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   2.7855153                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  9.2850511                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   4.6425255                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.7855153                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.9285051                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 3.7140204                  \n",
       "                                X410_HEK293T_P3B12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.8936426                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.5745705                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.3670533                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    0.9468213                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.4202320                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 5.2075172                  \n",
       "                                X411_HEK293T_P3C12_tetCTCF_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.1304493                      \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.3479207                      \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.6956740                      \n",
       "HEK293T_P1B6_SP4_d005_peak_3    0.5652247                      \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.5652247                      \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 0.5652247                      \n",
       "                                X412_HEK293T_P3D12_tetCTCF_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.8257871                      \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.9117439                      \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   0.6085957                      \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.4343827                      \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.2171914                      \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.8257871                      \n",
       "                                X413_HEK293T_P3E12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.7031904                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  9.1414749                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.8127615                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.4063807                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10         NA                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 0.7031904                  \n",
       "                                X414_HEK293T_P3F12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.0822624                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  6.8543286                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   0.3607541                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.1645248                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.4430165                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 4.3290496                  \n",
       "                                X415_HEK293T_P3G12_negctl_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   2.313741                      \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.519658                      \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.156870                      \n",
       "HEK293T_P1B6_SP4_d005_peak_3    4.627482                      \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.156870                      \n",
       "HEK293T_P3F10_SPI1_d025_peak_18       NA                      \n",
       "                                X416_HEK293T_P3H12_negctl_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.4081195                     \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.4487170                     \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   0.7040597                     \n",
       "HEK293T_P1B6_SP4_d005_peak_3    3.5202987                     \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.4081195                     \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 0.7040597                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>608176</li><li>252</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 608176\n",
       "\\item 252\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 608176\n",
       "2. 252\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 608176    252"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read in ATAC CPM values\n",
    "cpm_df <- read.csv(\"../../output/02-atac/01/cpm.tsv\", sep=\" \") # just use the cpm normalized peak count matrix\n",
    "cpm_df %>% head\n",
    "dim(cpm_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "174f1f43-d415-4c2a-b862-21b37728e519",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 608176 633 \n",
       "metadata(0):\n",
       "assays(3): motifScores motifMatches motifCounts\n",
       "rownames: NULL\n",
       "rowData names(0):\n",
       "colnames(633): MA0030.1_FOXF2 MA0031.1_FOXD1 ... MA0528.2_ZNF263\n",
       "  MA0609.2_CREM\n",
       "colData names(1): name"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-09-19 10:57:31     5.1  STATUS STARTED Updating fragment RDS file references\n",
      "2024-09-19 10:57:39     5.1  STATUS COMPLETED Updating fragment RDS file references\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# read the motif matching results\n",
    "motif.scores <- readRDS(\"../../output/02-atac/03/motif.scores.JASPAR2020.rds\") # this is JASPAR motifs\n",
    "motif.scores\n",
    "\n",
    "if (is.null(rownames(motif.scores))){\n",
    "    # read in the motif peak names\n",
    "    dsa_norm <- loadDsAcc('../../output/02-atac/01/dsa_norm_rpkmlog2quantile/')\n",
    "    region <- \"filteredConsensus\"\n",
    "\n",
    "    Peaks <- as.data.frame(dsa_norm@coord[[region]],row.names = NULL, stringsAsFactors=FALSE) %>%\n",
    "    dplyr::select(c(\"seqnames\",\"start\",\"end\",\"name\"))\n",
    "    rownames(motif.scores) <- Peaks$name\n",
    "    Peaks <- GRanges(Peaks)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40574b94-4438-42bc-8709-3ec4144834d2",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "### inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "230e0298-d962-45d4-98e6-cfafa87a71f7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# inputs\n",
    "motif_code_ls <- c(\n",
    "                   #\"ALX4\"=\"MA0681.2_PHOX2B\",\n",
    "                   #\"ELF1\"=\"MA0473.3_ELF1\",\n",
    "                   #\"IRF4\"=\"MA1419.1_IRF4\",\n",
    "                   \"KLF1\"=\"MA0039.4_KLF4\",\n",
    "                   \"KLF4\"=\"MA0039.4_KLF4\",\n",
    "                   #\"LEF1\"=\"MA0768.1_LEF1\",\n",
    "                   #\"NR4A1\"=\"MA1112.2_NR4A1\",\n",
    "                   #\"OCT4\"=\"MA1115.1_POU5F1\",\n",
    "                   #\"PRDM1\"=\"MA0508.3_PRDM1\",\n",
    "                   \"SOX2\"=\"MA0143.4_SOX2\",\n",
    "                   #\"SP4\"=\"MA0685.1_SP4\",\n",
    "                   \"SPI1\"=\"MA0080.5_SPI1\")\n",
    "                   #\"TCF3\"=\"MA0522.3_TCF3\")\n",
    "\n",
    "group1 = \"GFP_1\" # reference level\n",
    "region = \"filteredConsensus\"\n",
    "\n",
    "i = 3\n",
    "TF = names(motif_code_ls)[i]\n",
    "group2 = paste0(TF, \"_1\")\n",
    "motif_code = motif_code_ls[i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc8f45ac-cb73-4b1a-b418-97c3ff12a045",
   "metadata": {},
   "source": [
    "### read TF specific ATAC CPM, motif matching, differential analysis and sensitivity analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f8495950-1576-4a78-8a19-10717512ff55",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 18</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X81_HEK293T_P1A12_GFP_d100</th><th scope=col>X82_HEK293T_P1B12_GFP_d100</th><th scope=col>X245_HEK293T_P2A12_GFP_d100</th><th scope=col>X246_HEK293T_P2B12_GFP_d100</th><th scope=col>X409_HEK293T_P3A12_GFP_d100</th><th scope=col>X410_HEK293T_P3B12_GFP_d100</th><th scope=col>X413_HEK293T_P3E12_GFP_d100</th><th scope=col>X414_HEK293T_P3F12_GFP_d100</th><th scope=col>X369_HEK293T_P3A7_SOX2_d100</th><th scope=col>X370_HEK293T_P3B7_SOX2_d100</th><th scope=col>X377_HEK293T_P3A8_SOX2_d075</th><th scope=col>X378_HEK293T_P3B8_SOX2_d075</th><th scope=col>X385_HEK293T_P3A9_SOX2_d050</th><th scope=col>X386_HEK293T_P3B9_SOX2_d050</th><th scope=col>X393_HEK293T_P3A10_SOX2_d025</th><th scope=col>X394_HEK293T_P3B10_SOX2_d025</th><th scope=col>X401_HEK293T_P3A11_SOX2_d005</th><th scope=col>X402_HEK293T_P3B11_SOX2_d005</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>HEK293T_P3A12_GFP_d100_peak_1</th><td> 1.466495</td><td>1.566060</td><td>1.0934822</td><td>0.2761946</td><td>2.7855153</td><td>1.8936426</td><td>0.7031904</td><td>1.0822624</td><td>2.923978</td><td>1.367258</td><td>0.8990413</td><td>0.6822595</td><td>2.631582</td><td>0.4210129</td><td>1.4136354</td><td>2.679279</td><td>0.5821505</td><td>1.8272561</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3A12_GFP_d100_peak_17</th><td>10.632089</td><td>9.983635</td><td>6.5608929</td><td>4.1429185</td><td>9.2850511</td><td>7.5745705</td><td>9.1414749</td><td>6.8543286</td><td>8.187139</td><td>7.519918</td><td>8.9904126</td><td>8.8693732</td><td>7.142865</td><td>7.9992455</td><td>2.8272707</td><td>6.124067</td><td>4.0750532</td><td>6.6999390</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3G8_SPI1_d075_peak_6</th><td> 1.466495</td><td>1.566060</td><td>1.0934822</td><td>2.2095566</td><td>4.6425255</td><td>2.3670533</td><td>2.8127615</td><td>0.3607541</td><td>1.754387</td><td>2.050887</td><td>3.5961650</td><td>2.0467784</td><td>2.631582</td><td>3.7891163</td><td>1.4136354</td><td>1.913771</td><td>       NA</td><td>0.6090854</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P1B6_SP4_d005_peak_3</th><td> 2.199742</td><td>2.349091</td><td>3.2804465</td><td>2.2095566</td><td>2.7855153</td><td>0.9468213</td><td>1.4063807</td><td>2.1645248</td><td>1.754387</td><td>2.050887</td><td>1.7980825</td><td>2.0467784</td><td>1.879701</td><td>2.1050646</td><td>1.4136354</td><td>1.531017</td><td>0.5821505</td><td>0.6090854</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3G7_SPI1_d100_peak_10</th><td> 1.466495</td><td>1.957575</td><td>1.4579762</td><td>0.5523891</td><td>0.9285051</td><td>1.4202320</td><td>       NA</td><td>1.4430165</td><td>1.169591</td><td>      NA</td><td>0.8990413</td><td>2.0467784</td><td>1.127821</td><td>0.8420258</td><td>0.7068177</td><td>      NA</td><td>1.1643009</td><td>0.6090854</td></tr>\n",
       "\t<tr><th scope=row>HEK293T_P3F10_SPI1_d025_peak_18</th><td> 3.299614</td><td>2.544848</td><td>0.7289881</td><td>1.6571674</td><td>3.7140204</td><td>5.2075172</td><td>0.7031904</td><td>4.3290496</td><td>1.169591</td><td>2.734515</td><td>0.8990413</td><td>3.4112974</td><td>2.631582</td><td>2.1050646</td><td>1.4136354</td><td>1.531017</td><td>2.3286018</td><td>1.8272561</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 18\n",
       "\\begin{tabular}{r|llllllllllllllllll}\n",
       "  & X81\\_HEK293T\\_P1A12\\_GFP\\_d100 & X82\\_HEK293T\\_P1B12\\_GFP\\_d100 & X245\\_HEK293T\\_P2A12\\_GFP\\_d100 & X246\\_HEK293T\\_P2B12\\_GFP\\_d100 & X409\\_HEK293T\\_P3A12\\_GFP\\_d100 & X410\\_HEK293T\\_P3B12\\_GFP\\_d100 & X413\\_HEK293T\\_P3E12\\_GFP\\_d100 & X414\\_HEK293T\\_P3F12\\_GFP\\_d100 & X369\\_HEK293T\\_P3A7\\_SOX2\\_d100 & X370\\_HEK293T\\_P3B7\\_SOX2\\_d100 & X377\\_HEK293T\\_P3A8\\_SOX2\\_d075 & X378\\_HEK293T\\_P3B8\\_SOX2\\_d075 & X385\\_HEK293T\\_P3A9\\_SOX2\\_d050 & X386\\_HEK293T\\_P3B9\\_SOX2\\_d050 & X393\\_HEK293T\\_P3A10\\_SOX2\\_d025 & X394\\_HEK293T\\_P3B10\\_SOX2\\_d025 & X401\\_HEK293T\\_P3A11\\_SOX2\\_d005 & X402\\_HEK293T\\_P3B11\\_SOX2\\_d005\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tHEK293T\\_P3A12\\_GFP\\_d100\\_peak\\_1 &  1.466495 & 1.566060 & 1.0934822 & 0.2761946 & 2.7855153 & 1.8936426 & 0.7031904 & 1.0822624 & 2.923978 & 1.367258 & 0.8990413 & 0.6822595 & 2.631582 & 0.4210129 & 1.4136354 & 2.679279 & 0.5821505 & 1.8272561\\\\\n",
       "\tHEK293T\\_P3A12\\_GFP\\_d100\\_peak\\_17 & 10.632089 & 9.983635 & 6.5608929 & 4.1429185 & 9.2850511 & 7.5745705 & 9.1414749 & 6.8543286 & 8.187139 & 7.519918 & 8.9904126 & 8.8693732 & 7.142865 & 7.9992455 & 2.8272707 & 6.124067 & 4.0750532 & 6.6999390\\\\\n",
       "\tHEK293T\\_P3G8\\_SPI1\\_d075\\_peak\\_6 &  1.466495 & 1.566060 & 1.0934822 & 2.2095566 & 4.6425255 & 2.3670533 & 2.8127615 & 0.3607541 & 1.754387 & 2.050887 & 3.5961650 & 2.0467784 & 2.631582 & 3.7891163 & 1.4136354 & 1.913771 &        NA & 0.6090854\\\\\n",
       "\tHEK293T\\_P1B6\\_SP4\\_d005\\_peak\\_3 &  2.199742 & 2.349091 & 3.2804465 & 2.2095566 & 2.7855153 & 0.9468213 & 1.4063807 & 2.1645248 & 1.754387 & 2.050887 & 1.7980825 & 2.0467784 & 1.879701 & 2.1050646 & 1.4136354 & 1.531017 & 0.5821505 & 0.6090854\\\\\n",
       "\tHEK293T\\_P3G7\\_SPI1\\_d100\\_peak\\_10 &  1.466495 & 1.957575 & 1.4579762 & 0.5523891 & 0.9285051 & 1.4202320 &        NA & 1.4430165 & 1.169591 &       NA & 0.8990413 & 2.0467784 & 1.127821 & 0.8420258 & 0.7068177 &       NA & 1.1643009 & 0.6090854\\\\\n",
       "\tHEK293T\\_P3F10\\_SPI1\\_d025\\_peak\\_18 &  3.299614 & 2.544848 & 0.7289881 & 1.6571674 & 3.7140204 & 5.2075172 & 0.7031904 & 4.3290496 & 1.169591 & 2.734515 & 0.8990413 & 3.4112974 & 2.631582 & 2.1050646 & 1.4136354 & 1.531017 & 2.3286018 & 1.8272561\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 18\n",
       "\n",
       "| <!--/--> | X81_HEK293T_P1A12_GFP_d100 &lt;dbl&gt; | X82_HEK293T_P1B12_GFP_d100 &lt;dbl&gt; | X245_HEK293T_P2A12_GFP_d100 &lt;dbl&gt; | X246_HEK293T_P2B12_GFP_d100 &lt;dbl&gt; | X409_HEK293T_P3A12_GFP_d100 &lt;dbl&gt; | X410_HEK293T_P3B12_GFP_d100 &lt;dbl&gt; | X413_HEK293T_P3E12_GFP_d100 &lt;dbl&gt; | X414_HEK293T_P3F12_GFP_d100 &lt;dbl&gt; | X369_HEK293T_P3A7_SOX2_d100 &lt;dbl&gt; | X370_HEK293T_P3B7_SOX2_d100 &lt;dbl&gt; | X377_HEK293T_P3A8_SOX2_d075 &lt;dbl&gt; | X378_HEK293T_P3B8_SOX2_d075 &lt;dbl&gt; | X385_HEK293T_P3A9_SOX2_d050 &lt;dbl&gt; | X386_HEK293T_P3B9_SOX2_d050 &lt;dbl&gt; | X393_HEK293T_P3A10_SOX2_d025 &lt;dbl&gt; | X394_HEK293T_P3B10_SOX2_d025 &lt;dbl&gt; | X401_HEK293T_P3A11_SOX2_d005 &lt;dbl&gt; | X402_HEK293T_P3B11_SOX2_d005 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HEK293T_P3A12_GFP_d100_peak_1 |  1.466495 | 1.566060 | 1.0934822 | 0.2761946 | 2.7855153 | 1.8936426 | 0.7031904 | 1.0822624 | 2.923978 | 1.367258 | 0.8990413 | 0.6822595 | 2.631582 | 0.4210129 | 1.4136354 | 2.679279 | 0.5821505 | 1.8272561 |\n",
       "| HEK293T_P3A12_GFP_d100_peak_17 | 10.632089 | 9.983635 | 6.5608929 | 4.1429185 | 9.2850511 | 7.5745705 | 9.1414749 | 6.8543286 | 8.187139 | 7.519918 | 8.9904126 | 8.8693732 | 7.142865 | 7.9992455 | 2.8272707 | 6.124067 | 4.0750532 | 6.6999390 |\n",
       "| HEK293T_P3G8_SPI1_d075_peak_6 |  1.466495 | 1.566060 | 1.0934822 | 2.2095566 | 4.6425255 | 2.3670533 | 2.8127615 | 0.3607541 | 1.754387 | 2.050887 | 3.5961650 | 2.0467784 | 2.631582 | 3.7891163 | 1.4136354 | 1.913771 |        NA | 0.6090854 |\n",
       "| HEK293T_P1B6_SP4_d005_peak_3 |  2.199742 | 2.349091 | 3.2804465 | 2.2095566 | 2.7855153 | 0.9468213 | 1.4063807 | 2.1645248 | 1.754387 | 2.050887 | 1.7980825 | 2.0467784 | 1.879701 | 2.1050646 | 1.4136354 | 1.531017 | 0.5821505 | 0.6090854 |\n",
       "| HEK293T_P3G7_SPI1_d100_peak_10 |  1.466495 | 1.957575 | 1.4579762 | 0.5523891 | 0.9285051 | 1.4202320 |        NA | 1.4430165 | 1.169591 |       NA | 0.8990413 | 2.0467784 | 1.127821 | 0.8420258 | 0.7068177 |       NA | 1.1643009 | 0.6090854 |\n",
       "| HEK293T_P3F10_SPI1_d025_peak_18 |  3.299614 | 2.544848 | 0.7289881 | 1.6571674 | 3.7140204 | 5.2075172 | 0.7031904 | 4.3290496 | 1.169591 | 2.734515 | 0.8990413 | 3.4112974 | 2.631582 | 2.1050646 | 1.4136354 | 1.531017 | 2.3286018 | 1.8272561 |\n",
       "\n"
      ],
      "text/plain": [
       "                                X81_HEK293T_P1A12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1    1.466495                 \n",
       "HEK293T_P3A12_GFP_d100_peak_17  10.632089                 \n",
       "HEK293T_P3G8_SPI1_d075_peak_6    1.466495                 \n",
       "HEK293T_P1B6_SP4_d005_peak_3     2.199742                 \n",
       "HEK293T_P3G7_SPI1_d100_peak_10   1.466495                 \n",
       "HEK293T_P3F10_SPI1_d025_peak_18  3.299614                 \n",
       "                                X82_HEK293T_P1B12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.566060                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  9.983635                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.566060                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.349091                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.957575                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.544848                  \n",
       "                                X245_HEK293T_P2A12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.0934822                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  6.5608929                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.0934822                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    3.2804465                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.4579762                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 0.7289881                  \n",
       "                                X246_HEK293T_P2B12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.2761946                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  4.1429185                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.2095566                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.2095566                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.5523891                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.6571674                  \n",
       "                                X409_HEK293T_P3A12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   2.7855153                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  9.2850511                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   4.6425255                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.7855153                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.9285051                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 3.7140204                  \n",
       "                                X410_HEK293T_P3B12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.8936426                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.5745705                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.3670533                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    0.9468213                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.4202320                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 5.2075172                  \n",
       "                                X413_HEK293T_P3E12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.7031904                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  9.1414749                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.8127615                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.4063807                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10         NA                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 0.7031904                  \n",
       "                                X414_HEK293T_P3F12_GFP_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.0822624                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  6.8543286                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   0.3607541                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.1645248                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.4430165                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 4.3290496                  \n",
       "                                X369_HEK293T_P3A7_SOX2_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   2.923978                   \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.187139                   \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.754387                   \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.754387                   \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.169591                   \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.169591                   \n",
       "                                X370_HEK293T_P3B7_SOX2_d100\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.367258                   \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.519918                   \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.050887                   \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.050887                   \n",
       "HEK293T_P3G7_SPI1_d100_peak_10        NA                   \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.734515                   \n",
       "                                X377_HEK293T_P3A8_SOX2_d075\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.8990413                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.9904126                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   3.5961650                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.7980825                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.8990413                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 0.8990413                  \n",
       "                                X378_HEK293T_P3B8_SOX2_d075\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.6822595                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  8.8693732                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.0467784                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.0467784                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  2.0467784                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 3.4112974                  \n",
       "                                X385_HEK293T_P3A9_SOX2_d050\n",
       "HEK293T_P3A12_GFP_d100_peak_1   2.631582                   \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.142865                   \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   2.631582                   \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.879701                   \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.127821                   \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.631582                   \n",
       "                                X386_HEK293T_P3B9_SOX2_d050\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.4210129                  \n",
       "HEK293T_P3A12_GFP_d100_peak_17  7.9992455                  \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   3.7891163                  \n",
       "HEK293T_P1B6_SP4_d005_peak_3    2.1050646                  \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.8420258                  \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.1050646                  \n",
       "                                X393_HEK293T_P3A10_SOX2_d025\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.4136354                   \n",
       "HEK293T_P3A12_GFP_d100_peak_17  2.8272707                   \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.4136354                   \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.4136354                   \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.7068177                   \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.4136354                   \n",
       "                                X394_HEK293T_P3B10_SOX2_d025\n",
       "HEK293T_P3A12_GFP_d100_peak_1   2.679279                    \n",
       "HEK293T_P3A12_GFP_d100_peak_17  6.124067                    \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   1.913771                    \n",
       "HEK293T_P1B6_SP4_d005_peak_3    1.531017                    \n",
       "HEK293T_P3G7_SPI1_d100_peak_10        NA                    \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.531017                    \n",
       "                                X401_HEK293T_P3A11_SOX2_d005\n",
       "HEK293T_P3A12_GFP_d100_peak_1   0.5821505                   \n",
       "HEK293T_P3A12_GFP_d100_peak_17  4.0750532                   \n",
       "HEK293T_P3G8_SPI1_d075_peak_6          NA                   \n",
       "HEK293T_P1B6_SP4_d005_peak_3    0.5821505                   \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  1.1643009                   \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 2.3286018                   \n",
       "                                X402_HEK293T_P3B11_SOX2_d005\n",
       "HEK293T_P3A12_GFP_d100_peak_1   1.8272561                   \n",
       "HEK293T_P3A12_GFP_d100_peak_17  6.6999390                   \n",
       "HEK293T_P3G8_SPI1_d075_peak_6   0.6090854                   \n",
       "HEK293T_P1B6_SP4_d005_peak_3    0.6090854                   \n",
       "HEK293T_P3G7_SPI1_d100_peak_10  0.6090854                   \n",
       "HEK293T_P3F10_SPI1_d025_peak_18 1.8272561                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>608176</li><li>18</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 608176\n",
       "\\item 18\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 608176\n",
       "2. 18\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 608176     18"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cpm <- cpm_df[,c(grep(\"GFP\", colnames(cpm_df)), grep(TF, colnames(cpm_df)))]\n",
    "cpm %>% head\n",
    "dim(cpm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be644eac-a568-4ad4-a10b-9e73420e7e93",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 37251 1 \n",
       "metadata(0):\n",
       "assays(3): motifScores motifMatches motifCounts\n",
       "rownames(37251): HEK293T_P3G3_KLF1_d075_peak_117\n",
       "  HEK293T_P3H2_KLF1_d100_peak_135 ...\n",
       "  HEK293T_P3B2_OCT4_d100_peak_331409\n",
       "  HEK293T_P3F10_SPI1_d025_peak_1074237\n",
       "rowData names(6): sensitive saturating ... state group\n",
       "colnames(1): MA0143.4_SOX2\n",
       "colData names(1): name"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# identify peaks containing target motif\n",
    "target_motif_counts <- rowSums(motifCounts(motif.scores[,motif_code]))\n",
    "subset <- motif.scores[(target_motif_counts > 0), motif_code]\n",
    "meta <- read.table(paste0(\"../../output/02-atac/10/\", TF,\"_motif_containing_peak_meta.tsv\"), row.names=1, sep=\"\\t\")\n",
    "rowData(subset) <- meta[rownames(subset),]\n",
    "subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a86d8fae-d553-4ce2-a1a5-0f4e4ffb7ef5",
   "metadata": {
    "tags": []
   },
   "source": [
    "### compare median ATAC CPM for highest dose sample in different peak groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "22df6e33-f179-4176-a29c-778295742a21",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1068 rows containing non-finite outside the scale range (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 24 rows containing non-finite outside the scale range (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 29376 rows containing non-finite outside the scale range\n",
      "(`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 30468 rows containing non-finite outside the scale range\n",
      "(`stat_boxplot()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 30468 rows containing non-finite outside the scale range\n",
      "(`stat_boxplot()`).”\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "\n",
       "[[2]]\n",
       "\n",
       "[[3]]\n",
       "\n",
       "[[4]]\n",
       "\n",
       "[[5]]\n",
       "\n",
       "[[6]]\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "maxdose_cols <- colnames(cpm)[grep(\"100\", colnames(cpm))]\n",
    "idx <- which(rowData(subset)$saturating==\"Yes\")\n",
    "means1 <- cpm[rownames(subset)[idx],maxdose_cols] %>% rowMeans\n",
    "f1 <- means1 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at highest plasmid dose \\nsaturating sensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means1),2)))\n",
    "\n",
    "idx <- which(rowData(subset)$saturating==\"No\")\n",
    "means2 <- cpm[rownames(subset)[idx],maxdose_cols] %>% rowMeans\n",
    "f2 <- means2 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at highest plasmid dose \\nnonsaturating sensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means2),2)))\n",
    "\n",
    "idx <- which(rowData(subset)$state==\"open\")\n",
    "means3 <- cpm[rownames(subset)[idx],maxdose_cols] %>% rowMeans\n",
    "f3 <- means3 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at highest plasmid dose \\nopen nonsensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means3),2)))\n",
    "\n",
    "idx <- which(rowData(subset)$state==\"closed\")\n",
    "means4 <- cpm[rownames(subset)[idx],maxdose_cols] %>% rowMeans\n",
    "f4 <- means4 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at highest plasmid dose \\nclosed nonsensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means4),2)))\n",
    "\n",
    "df <- c(means1,means2,means3,means4) %>% as.data.frame\n",
    "colnames(df) <- \"ATAC_CPM\"\n",
    "df$group <- c(rep(\"saturating sensitive\", length(means1)),\n",
    "             rep(\"nonsaturating sensitive\", length(means2)),\n",
    "             rep(\"open nonsensitive\", length(means3)),\n",
    "             rep(\"closed nonsensitive\", length(means4)))\n",
    "f5 <- ggboxplot(df, x=\"group\", y=\"ATAC_CPM\") + ggtitle(paste0(TF, \" ATAC CPM at highest TF dose\")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "f6 <- ggboxplot(df, x=\"group\", y=\"ATAC_CPM\") + ggtitle(paste0(TF, \" ATAC CPM at highest TF dose\")) + coord_cartesian(ylim = c(-1, 10)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "\n",
    "\n",
    "pdf(paste0(plotdir, \"/\",TF,\"_\",motif_code,\"_ATAC_CPM_highdose.pdf\"), width=5, height=5)\n",
    "lapply(paste0(\"f\",1:6),function(n){get(n)})\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c0ae370-9fb9-4f6a-8328-6f6b9a654a70",
   "metadata": {
    "tags": []
   },
   "source": [
    "### compare median ATAC CPM for control sample in different peak groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a0dd7f8f-b882-42ed-9ef8-6ec633471624",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1068 rows containing non-finite outside the scale range (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 24 rows containing non-finite outside the scale range (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 27532 rows containing non-finite outside the scale range\n",
      "(`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 28624 rows containing non-finite outside the scale range\n",
      "(`stat_boxplot()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 28624 rows containing non-finite outside the scale range\n",
      "(`stat_boxplot()`).”\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "\n",
       "[[2]]\n",
       "\n",
       "[[3]]\n",
       "\n",
       "[[4]]\n",
       "\n",
       "[[5]]\n",
       "\n",
       "[[6]]\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mindose_cols <- colnames(cpm)[grep(\"GFP\", colnames(cpm))]\n",
    "idx <- which(rowData(subset)$saturating==\"Yes\")\n",
    "means1 <- cpm[rownames(subset)[idx],mindose_cols] %>% rowMeans\n",
    "f1 <- means1 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at 0 plasmid dose \\nsaturating sensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means1),2)))\n",
    "\n",
    "idx <- which(rowData(subset)$saturating==\"No\")\n",
    "means2 <- cpm[rownames(subset)[idx],mindose_cols] %>% rowMeans\n",
    "f2 <- means2 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at 0 plasmid dose \\nnonsaturating sensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means2),2)))\n",
    "\n",
    "idx <- which(rowData(subset)$state==\"open\")\n",
    "means3 <- cpm[rownames(subset)[idx],mindose_cols] %>% rowMeans\n",
    "f3 <- means3 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at 0 plasmid dose \\nopen nonsensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means3),2)))\n",
    "\n",
    "idx <- which(rowData(subset)$state==\"closed\")\n",
    "means4 <- cpm[rownames(subset)[idx],mindose_cols] %>% rowMeans\n",
    "f4 <- means4 %>% gghistogram + \n",
    "    ggtitle(sprintf(\"distribution of ATAC CPM at 0 plasmid dose \\nclosed nonsensitive peaks \\n(n=%s, median=%s)\", length(idx), round(median(means4),2)))\n",
    "\n",
    "df <- c(means1,means2,means3,means4) %>% as.data.frame\n",
    "colnames(df) <- \"ATAC_CPM\"\n",
    "df$group <- c(rep(\"saturating sensitive\", length(means1)),\n",
    "             rep(\"nonsaturating sensitive\", length(means2)),\n",
    "             rep(\"open nonsensitive\", length(means3)),\n",
    "             rep(\"closed nonsensitive\", length(means4)))\n",
    "f5 <- ggboxplot(df, x=\"group\", y=\"ATAC_CPM\") + ggtitle(paste0(TF, \" ATAC CPM at 0 TF dose\")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "f6 <- ggboxplot(df, x=\"group\", y=\"ATAC_CPM\") + ggtitle(paste0(TF, \" ATAC CPM at 0 TF dose\")) + coord_cartesian(ylim = c(-1, 10)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "\n",
    "pdf(paste0(plotdir, \"/\", TF,\"_\",motif_code,\"_ATAC_CPM_nodose.pdf\"), width=5, height=5)\n",
    "lapply(paste0(\"f\",1:6),function(n){get(n)})\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7ead01-d3c6-4ac0-a122-59476cb98578",
   "metadata": {},
   "source": [
    "### ChIP overlap with peak groups\n",
    "download from ENCODE, list of ChIP-seq datasets are in 17c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9e488922-c409-4c26-aadb-69f50cee193a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plot_overlap_chip <- function(bed, overlap, name){\n",
    "    nhit <- length(overlap)\n",
    "    bedsub <- bed[overlap@to]\n",
    "    medsig <- bedsub$signalValue %>% median %>% round(2)\n",
    "    medq <- bedsub$qValue %>% median %>% round(2)\n",
    "\n",
    "    pp1 <- bedsub$signalValue %>% gghistogram + \n",
    "        ggtitle(paste0(mod, \" ChIP signalValue distribution\\n\", name, \" peaks \\nnoverlap=\", nhit,\" median=\", medsig)) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    pp2 <- bedsub$qValue %>% gghistogram + \n",
    "        ggtitle(paste0(mod, \" ChIP -log10(qValue) distribution\\n\", name, \" peaks\\nnoverlap=\", nhit,\" median=\", medq)) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    return(list(pp1,pp2))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df941cba-8b52-437c-9d94-d23ac1b7b6b0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n",
      "Warning message:\n",
      "“Using `bins = 30` by default. Pick better value with the argument `bins`.”\n"
     ]
    }
   ],
   "source": [
    "# single block for easy running\n",
    "sample <- read.csv(\"17b_encode_sample_info.csv\")\n",
    "\n",
    "for (mod in c(\"H3K27ac\", \"H3K9me3\", \"H3K4me1\", \"H3K4me3\", \"H3K36me3-1\", \"H3K36me3-2\")){\n",
    "    bed <- read.table(paste0(outdir, \"/\",sample[sample$mod==mod, \"bedfile\"]), sep=\"\\t\")\n",
    "    colnames(bed) <- c(\"chrom\", \"chromStart\", \"chromEnd\", \"name\",\"score\",\"strand\",\"signalValue\",\"pValue\",\"qValue\",\"peak\")\n",
    "    bed <- GRanges(bed)\n",
    "    # macs2 narrow peaks output: signal value - measurement of overall enrichment for the region, \n",
    "    #                            pValue - statistical significance (-log10), qValue - Statistical significance using FDR (-log10), \n",
    "    #                            peak - Point-source called for this peak; 0-based offset from chromSTart\n",
    "    \n",
    "    plots <- list()\n",
    "    \n",
    "    # overlap each group of peaks with ChIP signal\n",
    "    query1 <- subset@rowRanges[which(rowData(subset)$saturating==\"Yes\")]\n",
    "    overlap1 <- findOverlaps(query1, bed)\n",
    "    plots <- c(plots, plot_overlap_chip(bed, overlap1, \"saturating sensitive\"))\n",
    "    \n",
    "    query2 <- subset@rowRanges[which(rowData(subset)$saturating==\"No\")]\n",
    "    overlap2 <- findOverlaps(query2, bed)\n",
    "    plots <- c(plots, plot_overlap_chip(bed, overlap2, \"nonsaturating sensitive\"))\n",
    "    \n",
    "    query3 <- subset@rowRanges[which(rowData(subset)$state==\"open\")]\n",
    "    overlap3 <- findOverlaps(query3, bed)\n",
    "    plots <- c(plots, plot_overlap_chip(bed, overlap3, \"open nonsensitive\"))\n",
    "    \n",
    "    query4 <- subset@rowRanges[which(rowData(subset)$state==\"closed\")]\n",
    "    overlap4 <- findOverlaps(query4, bed)\n",
    "    plots <- c(plots, plot_overlap_chip(bed, overlap4, \"closed nonsensitive\"))\n",
    "    \n",
    "    df <- c(bed[overlap1@to]$signalValue,bed[overlap2@to]$signalValue,bed[overlap3@to]$signalValue,bed[overlap4@to]$signalValue) %>% data.frame\n",
    "    colnames(df) <- \"signalValue\"\n",
    "    df$qValue <- c(bed[overlap1@to]$qValue,bed[overlap2@to]$qValue,bed[overlap3@to]$qValue, bed[overlap4@to]$qValue)\n",
    "    df$group <- c(rep(\"saturating_sensitive\", length(bed[overlap1@to]$signalValue)),\n",
    "                 rep(\"nonsaturating_sensitive\", length(bed[overlap2@to]$signalValue)),\n",
    "                 rep(\"open_nonsensitive\", length(bed[overlap3@to]$signalValue)),\n",
    "                 rep(\"closed_nonsensitive\", length(bed[overlap4@to]$signalValue)))\n",
    "    df$peakname <- c(names(query1[overlap1@from]), names(query2[overlap2@from]), names(query3[overlap3@from]), names(query4[overlap4@from]))\n",
    "\n",
    "    p9 <- ggplot(df, aes(x = group)) + geom_bar(stat = \"count\") + theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab(\"# hits\")\n",
    "    \n",
    "    \n",
    "    # plot percentage of peaks with the histone mod in each group\n",
    "    groups <- c(\"saturating_sensitive\", \"nonsaturating sensitive\", \"open_nonsensitive\", \"closed_nonsensitive\")\n",
    "    pct <- data.frame(group=factor(groups, levels=groups),\n",
    "                     total_peaks=c(length(query1),length(query2), length(query3), length(query4)),\n",
    "                      mod_peaks=c(length(unique(overlap1@from)),length(unique(overlap2@from)),\n",
    "                                 length(unique(overlap3@from)),length(unique(overlap4@from))))\n",
    "    pct$pct_mod_peaks <- round(pct$mod_peaks / pct$total_peaks * 100, 2)\n",
    "    p10 <- ggbarplot(pct, x=\"group\",y=\"pct_mod_peaks\",label = T, color=\"darkgray\",fill = \"darkgray\") + ylab(paste0(\"% of peaks with \",mod)) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    \n",
    "    # plot number of peaks per group\n",
    "    p11 <- ggbarplot(pct, x=\"group\",y=\"total_peaks\",label = T, color=\"darkgray\",fill = \"darkgray\") + ylab(\"# peaks in group\") + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    \n",
    "    # cross group comparison\n",
    "    p12 <- ggboxplot(df,x=\"group\",y=\"signalValue\") + ylab(paste0(mod, \" ChIP signalValue\")) + xlab(paste0(motif_code, \" motif containing peaks in \", TF, \" samples\")) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    p13 <- ggboxplot(df,x=\"group\",y=\"signalValue\") + ylab(paste0(mod, \" ChIP signalValue\")) + xlab(paste0(motif_code, \" motif containing peaks in \", TF, \" samples\")) + coord_cartesian(ylim = c(-1, 30)) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    \n",
    "    p14 <- ggboxplot(df,x=\"group\",y=\"qValue\") + ylab(paste0(mod, \" ChIP -log10(qValue)\")) + xlab(paste0(motif_code, \" motif containing peaks in \", TF, \" samples\")) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    p15 <- ggboxplot(df,x=\"group\",y=\"qValue\") + ylab(paste0(mod, \" ChIP -log10(qValue)\")) + xlab(paste0(motif_code, \" motif containing peaks in \", TF, \" samples\")) + coord_cartesian(ylim = c(-1, 50)) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "\n",
    "    # just comparing saturating vs nonsaturating sensitive peaks\n",
    "    dff <- df[df$group %in% c(\"saturating_sensitive\",\"nonsaturating_sensitive\"),]\n",
    "    p16 <- ggboxplot(dff,x=\"group\",y=\"signalValue\") + ylab(paste0(mod, \" ChIP signalValue\")) + xlab(paste0(motif_code, \" motif containing peaks in \", TF, \" samples\")) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    p17 <- ggboxplot(dff,x=\"group\",y=\"signalValue\") + ylab(paste0(mod, \" ChIP signalValue\")) + xlab(paste0(motif_code, \" motif containing peaks in \", TF, \" samples\")) + coord_cartesian(ylim = c(-1, 20)) + \n",
    "        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "    \n",
    "    # correlating EC50 and hill coef of saturating sensitive peaks with chip data\n",
    "    dff <- df[df$group==\"saturating_sensitive\",]\n",
    "    dff$EC50 <- rowData(subset)[dff$peakname, \"EC50\"]\n",
    "    dff$h <- rowData(subset)[dff$peakname, \"h\"]\n",
    "    \n",
    "    p18 <- ggplot(data=dff, aes(y=log10(EC50),x=log10(signalValue)))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "      geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "    p19 <- ggplot(data=dff, aes(y=log10(h),x=log10(signalValue)))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "          geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "    p20 <- ggplot(data=dff, aes(y=log10(EC50),x=log10(qValue)))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "          geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "    p21 <- ggplot(data=dff, aes(y=log10(h),x=log10(qValue)))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "          geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "\n",
    "    p22 <- ggplot(data=dff, aes(y=EC50,x=signalValue))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "      geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "    p23 <- ggplot(data=dff, aes(y=h,x=signalValue))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "          geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "    p24 <- ggplot(data=dff, aes(y=EC50,x=qValue))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "          geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "    p25 <- ggplot(data=dff, aes(y=h,x=qValue))+ geom_point() + theme(text = element_text(size=20)) +\n",
    "          geom_smooth(method='lm', formula= y~x) + stat_cor(aes(label =  ..rr.label..))\n",
    "\n",
    "\n",
    "    pdf(paste0(plotdir, \"/\", TF,\"_\",motif_code,\"_\",mod,\".pdf\"), width=5, height=5)\n",
    "    for (p in plots){print(p)}\n",
    "    lapply(paste0(\"p\",9:25),function(n){print(get(n))})\n",
    "    dev.off()\n",
    "}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
